lang/cs/Org.Apache.REEF.Wake/Avro/ProtocolSerializer.cs

```diff
@@ -0,0 +1,172 @@
+                {
+                    genericInfo = registerInfo.MakeGenericMethod(new[] { type });
+                    genericInfo.Invoke(this, null);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Generate and store the metadata necessary to serialze and deserialize a specific message type.
+        /// </summary>
+        /// <typeparam name="TMessage">The class type of the message being registered.</typeparam>
+        internal void Register<TMessage>()
+        {
+            Logr.Log(Level.Info, "Registering message type: {0} {1}", typeof(TMessage).FullName, typeof(TMessage).Name);
+
+            IAvroSerializer<TMessage> messageSerializer = AvroSerializer.Create<TMessage>();
+            Serialize serialize = (MemoryStream stream, object message) =>
+            {
+                messageSerializer.Serialize(stream, (TMessage)message);
+            };
+            serializeMap.Add(typeof(TMessage).Name, serialize);
+
+            Deserialize deserialize = (MemoryStream stream, object observer, long sequence) =>
+            {
+                TMessage message = messageSerializer.Deserialize(stream);
+                IObserver<MessageInstance<TMessage>> msgObserver = observer as IObserver<MessageInstance<TMessage>>;
+                if (msgObserver != null)
+                {
+                    msgObserver.OnNext(new MessageInstance<TMessage>(sequence, message));
+                }
+                else
+                {
+                    Logr.Log(Level.Warning, "Unhandled message received: {0}", message);
+                }
+            };
+            deserializeMap.Add(typeof(TMessage).Name, deserialize);
+        }
+
+        /// <summary>
+        /// Serialize the input message and return a byte array.
+        /// </summary>
+        /// <param name="message">An object reference to the messeage to be serialized</param>
+        /// <param name="sequence">A long which cotains the higher level protocols sequence number for the message.</param>
+        /// <returns>A byte array containing the serialized header and message.</returns>
+        public byte[] Write(object message, long sequence) 
+        {
+            Logr.Log(Level.Info, "Serializing message: {0}", message.GetType().Name);
+            try
+            { 
+                using (MemoryStream stream = new MemoryStream())
+                {
+                    Header header = new Header(sequence, message.GetType().Name);
+                    headerSerializer.Serialize(stream, header);
+
+                    Serialize serialize;
+                    string name = message.GetType().Name;
+                    if (serializeMap.TryGetValue(name, out serialize))
+                    {
+                        serialize(stream, message);
+                    }
+                    else
+                    {
+                        Logr.Log(Level.Info, "Request to seriaiize message of unknown type: {0}",  message.GetType().Name);
```

Comment:
Also, it would be nice to have symmetry with the `Read()` method below: either have log/`throw` pair in both cases or do just `throw` and log in the `catch` block.